app/services/foreman_bootdisk/iso_generator.rb

```diff
@@ -76,12 +88,35 @@ def self.generate(opts = {})
 
         if opts[:ipxe]
           ipxe_source_file = File.join(Setting[:bootdisk_ipxe_dir], 'ipxe.lkrn')
-          raise(Foreman::Exception, N_('Please ensure the ipxe-bootimgs package is installed.')) unless File.exist?(ipxe_source_file)
+          raise Foreman::Exception.new(N_('Please ensure the ipxe-bootimgs package is installed.')) unless File.exist?(ipxe_source_file)
 
           FileUtils.cp(ipxe_source_file, File.join(wd, 'build', 'ipxe'))
           File.open(File.join(wd, 'build', 'script'), 'w') { |file| file.write(opts[:ipxe]) }
         end
 
+        if opts[:grub]
+          grub_path = "/var/lib/tftpboot/grub2/grubx64.efi"
+          shim_path = "/var/lib/tftpboot/grub2/shimx64.efi"
+          raise Foreman::Exception.new(N_('Ensure %{path} exists by enabling TFTP smart proxy module.'), path: grub_path) unless File.exist?(grub_path)
+          raise Foreman::Exception.new(N_('Ensure %{path} exists by enabling TFTP smart proxy module.'), path: shim_path) unless File.exist?(shim_path)
+          FileUtils.mkdir_p(File.join(wd, 'build', 'EFI', 'BOOT'))
+          FileUtils.cp(grub_path, File.join(wd, 'build', 'EFI', 'BOOT', 'grubx64.efi'))
+          FileUtils.cp(shim_path, File.join(wd, 'build', 'EFI', 'BOOT', 'shimx64.efi'))
+          File.open(File.join(wd, 'build', 'EFI', 'BOOT', 'grub.cfg'), 'w') { |file| file.write(opts[:grub]) }
+          efibootimg = File.join(wd, 'build', 'efiboot.img')
+          system("mformat -T 7000 -C -i #{efibootimg}")
```

Comment:
I created a helper method and all system calls have error checking now. I tested `truncate` method and it works fine in libvirt, since ESP partitions are simple MSDOS FAT partitions this should work fine.